### PR TITLE
implement image versioning and publishing workflow in CI

### PR DESCRIPTION
**Pull request by edaniszewski**
implement the image versioning and deployment flow as described in https://docs.google.com/document/d/19ziK27Vp5BzE9OWcwFuDzZwPKrZYQb5s728VA3P4fRo/edit

note: this does not implement the monthly builds.

@timfallmk , since you are the resident CI expert, take a look through this and lemme know if it makes sense/if there are things that you think should change

### CODE REVIEW

**Comment by edaniszewski** [comment]
I've been playing around with some stuff to try and get things to a mostly generalized place so we can apply it to other repos as well. 

I moved some of the CI stuff to the makefile. Not totally sure I am a fan of that, but the way I see it, we can either:
- keep the makefile simple for local dev stuff only. this would mean that some stuff is re-defined in the CI config, so we would have to maintain information in two places (e.g. ldflags, names, etc)
- allow makefiles to do some of the heavy lifting for ci. this makes them a bit more complicated, but then we are maintaining a single source of info, not multiple. 

Something else I'm trying to figure out how to improve is the build and publish script. the script itself is okay, though may need some updates for some of our builds (e.g. for build-args). the other bit to this is that if we use the script, all repos will have a copy, so if it ever needs to get updated, it'll need to be updated everywhere which is a little bit of a pain.

File: .circleci/config.yml [hunk 1]
```diff
@@ -28,115 +52,121 @@ jobs:
...
       - run:
           name: Create Release
           command: |
+            go get -v github.com/tcnksm/ghr
             if git describe --exact-match --tags HEAD; then
               CIRCLE_TAG=$(git describe --exact-match --tags HEAD)
             fi
             ghr \
-              -u ${GITHUB_USER} \
+              -u ${CIRCLE_PROJECT_USERNAME} \
```
Comment (timfallmk) [inline .circleci/config.yml]: You _don't_ want all projects published under my name? How dare you!

File: Makefile [hunk 2]
```diff
@@ -2,32 +2,43 @@
...
 
 
+#
+# Local Targets
+#
+
 .PHONY: build
 build:  ## Build the plugin Go binary
-	go build -ldflags ${LDFLAGS} -o build/emulator
+	go build -ldflags "${LDFLAGS}" -o build/emulator
```
Comment (timfallmk) [inline Makefile]: I would double check make sure this run properly in CI. String escaping is especially sensitive in `ldflags`.

File: Makefile [hunk 3]
```diff
@@ -2,32 +2,43 @@
...
 
 .PHONY: build-linux
 build-linux:  ## Build the plugin for linux amd64
-	GOOS=linux GOARCH=amd64 go build -ldflags ${LDFLAGS} -o build/emulator .
+	GOOS=linux GOARCH=amd64 go build -ldflags "${LDFLAGS}" -o build/emulator .
```
Comment (timfallmk) [inline Makefile]: Ditto.

File: .circleci/config.yml (hunk 1 above)
Comment (edaniszewski) [reply .circleci/config.yml]: Oh. Maybe I'm just dumb. I think that I thought this was just the org name, e.g. vapor-ware.

Should probably change this back unless we actually have a "vapor-ware" user.. and if so, we'd need a token for it, right?

Comment (edaniszewski) [reply .circleci/config.yml]: we've got a `vaporauto` and `vaporio` user that could do this. OR to make things easier we could just stick with using your creds (:

File: Makefile (hunk 2 above)
Comment (edaniszewski) [reply Makefile]: It looks like this works:

```
#!/bin/bash -eo pipefail
make build

go build -ldflags "-w -X main.BuildDate=2018-05-11T16:52:00 -X main.GitCommit=5f25594 -X main.GitTag=1.0-6-g5f25594 -X main.GoVersion=go1.10.2 -X main.VersionString=1.0.1" -o build/emulator

```

**Pull request merged**